README.rst
``chacractl``
-------------
A client interface to interact with the ``chacra`` service
(http://github.com/ceph/chacra).


configuration
-------------
The client writes a configuration file at ``$HOME/.chacractl`` with some needed
information::

    # This file was automatically generated by the chacractl CLI
    # make sure to update it with the correct user and key to talk to the API

    url = "http://example/"
    user = "admin"
    key = "secret"
    ssl_verify = True

These values need to be modified so that the CLI can interact with the remote
API.

binaries
--------
The CLI allows for a few interactions with binaries, depending on the action
needed the requirements may change. For example when trying to upload a binary,
the tool can accept either ``stdin`` or an actual file input.

create
^^^^^^
For POSTing a binary, if posting a single binary it is best to just pass the
full path as the last argument to the ``binary`` subcommand::

    chacractl binary create project/ref/distro/distro-version/arch /path/to/binary.rpm

But for convenience, if sending many files, it can also accept them from stdin,
usually with the output of a ``find`` command that can be filtered with
``grep`` or similar.

Below is an example of such a command for a project like ``ceph-deploy``::

    find ~/repos | grep ceph-deploy | grep rpm | grep noarch | grep el6
    | chacractl binary create ceph-deploy/master/centos/6/noarch


If a binary already exists and there is a need to re-upload it, the ``--force``
flag must be used. In that case the request will be a ``PUT`` and the resource
will be overwritten in the API.

exists
------
Check whether a given url endpoint exists or not. This subcommand requires the
full url part (as opposed to other subcommands that require the url starting
from the project name)::

    chacractl exists repos/ceph-deploy/master/centos/7

If the endpoint is not found it will return a non-zero exit status


repos
-----
Repositories can be updated or fully re-created. This is useful when the API
changes and the repository creation wasn't triggered.

recreate
^^^^^^^^
A ``recreate`` call will fully remove a repository on disc and will instruct
the API to queue a repo creation. These actions are not immediate: the removal
of the repository will happen as soon as the request is processed and the
repository creation will be done when the queue does the job (possibly after
a configured timeout)

A simple call would look like::

    chacractl repo recreate ceph/infernalis/debian/jessie

Note how the full path to the distribution version is needed.

update
^^^^^^
Similar to ``recreate``, but it will not remove the actual repository on disc
before queuing the request. It is useful when a binary has been added and the
client wants to ensure it is added to the repo::

    chacractl repo update ceph/infernalis/debian/jessie

set-raw
^^^^^^^
Repos have a "type" inferred from the binaries they contain; if rpms
are uploaded, the repo will be created as an rpm repo; if debs, a Debian
repo will result.  'Raw' repos are simple containers for binaries that
need no further processing into repositories, such as .zip files or
other binary products that will be simply downloaded.  For such repos,
use set-raw to set the type of the repo so that no asynchronous repo
construction is requested.

    chacractl repo set-raw ceph/infernalis/windows/1809


CHANGELOG.rst
0.0.24
------
Release Date: Mar 1, 2022

* Add 'set-raw' command to support raw repos.  includes doc.

0.0.23
------
Release Date: Feb 16, 2021

* api: should not append "delete" in the url for delete

0.0.22
------
Release Date: Aug 20, 2020

* Add ``repo delete`` capability

0.0.21
------
Release Date: May 12, 2020

* Add Python3 compatibility

0.0.20
------
Release Date: January 30th, 2019

* Pinning ``requests-toolbelt`` to 0.9.1.  Prevents requiring ``pyopenssl``

0.0.19
------
Release Date: January 29th, 2019

* Add ``pyopenssl`` as a requirement since ``requests_toolbelt`` doesn't seem
  to require it anymore

0.0.18
------
Release Date: January 23th, 2019

* Release 0.0.17 did not contain the multipart upload support

0.0.17
------
**BOTCHED RELEASE**
Release Date: January 23th, 2019

* Use multipart uploads to prevent the 2GB issue that was supposedly fixed
  by 0.0.16

0.0.16
------
Release Date: August 20th, 2018

* Fix an issue for POSTing files larger than 2GB

0.0.15
------
Release Date: April 3rd, 2017

* Fix an issue where an HTTP response might be ``None``

0.0.14
------
Release Date: March 22nd, 2017

* Retry failed HTTP operations, improving robustness of the many requests
  the tool has to perform when dealing with large amounts of packages.


0.0.13
------
Release Date: December 9th, 2016

* Fix an issue with filenames vs. paths if a digest doesn't match


0.0.12
------
Release Date: September 28th, 2016

* Improve logging around checksum mismatches


0.0.12
------
Release Date: September 28th, 2016

* Improve logging around checksum mismatches


0.0.11
------
Release Date: September 27th, 2016

* Fix updating with checksum verification, where the URL would be wrong for PUT
  requests


0.0.10
------
Release Date: September 20th, 2016

* Fix ``TypeError`` when uploading. Missing ``self`` on method's signature


0.0.9
-----
Release Date: September 20th, 2016

* Add option to verify binaries were correctly uploaded via checksums.
* When uploading, fix a bug where a non-200 response would trigger an
  UnboundLocal error


0.0.8
-----
Release Date: June 13th, 2016

* Adds ``ssl_verify`` option for configuraton, to prevent issues on local
  chacra instances where self signed ssl certs are used
* Reset the chacractl config for tests that needs them
* add travis.yml configuration
* binaries: fix issue when deleting, treating 204 as an error
* binaries: silence warnings when deleting


0.0.7
-----
Release Date: December 21st, 2015

* Fix unicode issues when uploading binaries as text.
* Implement support for project creation


0.0.6
-----
Release Date: December 2nd, 2015

* Minor fix for DELETE operations, actually making them work

0.0.5
-----
Release Date: November 9rd, 2015

* Use the filename when doing a PUT to re-upload a binary

0.0.4
-----
Release Date: November 3rd, 2015

* Fix an issue when raising errors when posting to a 404 url

0.0.3
-----
Release Date: November 3rd, 2015

* Add a ``--force`` flag to be able to re-upload binaries

0.0.2
-----
Release Date: October 27th, 2015

* Implement the 'exists' subcommand to check existance of URL endpoints

0.0.1
-----
Release Date: October 20th, 2015

* Initial release.


chacractl/api/binaries.py
import logging
import sys
import os
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO
from textwrap import dedent
from hashlib import sha512

import requests
from requests_toolbelt.multipart.encoder import MultipartEncoder

from tambo import Transport

import chacractl
from chacractl.util import retry

logger = logging.getLogger(__name__)


class Binary(object):
    _help = dedent("""
    Operate binaries on a remote chacra instance.

    Creating a new binary::

        chacractl binary create project/ref/distro/distro_version/arch /path/to/binary

    Options:

    create        Creates a new binary at a given distro version architecture
    delete        Deletes an existing binary from chacra
    --force       If the resource exists, force the upload
    """)
    help_menu = "create, update metadata, or delete binaries"
    options = ['create', '--force', 'delete']

    def __init__(self, argv):
        self.argv = argv

    @property
    def base_url(self):
        return os.path.join(
            chacractl.config['url'], 'binaries'
        )

    def sanitize_filename(self, line):
        """
        lines may come with newlines and leading slashes make sure
        they are clean so that they can be processed
        """
        line = line.strip('\n')
        if os.path.isfile(line):
            return os.path.abspath(line)

    def sanitize_url(self, url_part):
        # get rid of the leading slash to prevent issues when joining
        url = url_part.lstrip('/')

        # and add a trailing slash so that the request is done at the correct
        # canonical url
        if not url.endswith('/'):
            url = "%s/" % url
        return url

    def get_checksum(self, filepath):
        chsum = sha512()
        with open(filepath, 'rb') as binary:
            for chunk in iter(lambda: binary.read(4096), b''):
                chsum.update(chunk)
        return chsum.hexdigest()

    @retry()
    def upload_is_verified(self, arch_url, filename, digest):
        r = requests.get(arch_url, verify=chacractl.config['ssl_verify'])
        r.raise_for_status()
        arch_data = r.json()
        remote_digest = arch_data[filename]['checksum']
        verified = remote_digest == digest
        if not verified:
            logging.error(
                    'Checksum mismatch: server has wrong checksum for %s',
                    filename)
            logging.error('local checksum: %s', digest)
            logging.error('remote checksum: %s', remote_digest)
        return verified

    @retry()
    def post(self, url, filepath):
        filename = os.path.basename(filepath)
        file_url = os.path.join(url, filename) + '/'
        exists = requests.head(file_url, verify=chacractl.config['ssl_verify'])
        digest = self.get_checksum(filepath)

        if exists.status_code == 200:
            if not self.force:
                logger.warning(
                    'resource exists and --force was not used, will not upload'
                )
                logger.warning('SKIP %s', file_url)
                return
            return self.put(file_url, filepath)
        elif exists.status_code == 404:
            length = os.path.getsize(filepath)
            logger.info('POSTing file: %s', filepath)
            mpart = MultipartEncoder(
                fields={'file': (filename, open(filepath, 'rb'), 'text/plain')}
            )

            response = requests.post(
                url,
                data=mpart,
                headers={
                    'Content-Type': mpart.content_type,
                    'Content-Length': '%d' % length,
                },
                auth=chacractl.config['credentials'],
                verify=chacractl.config['ssl_verify'])
            if response.status_code > 201:
                logger.warning("%s -> %s", response.status_code, response.text)
                response.raise_for_status()
        if not self.upload_is_verified(url, filename, digest):
            # Since this is a new file, attempt to delete it
            logging.error('Deleting corrupted file from server...')
            self.delete(file_url)
            raise SystemExit(
                    'Checksum mismatch: remote server has wrong checksum for %s'
                    % filepath)
    @retry()
    def put(self, url, filepath):
        filename = os.path.basename(filepath)
        logger.info('resource exists and --force was used, will re-upload')
        logger.info('PUTing file: %s', filepath)
        digest = self.get_checksum(filepath)
        length = os.path.getsize(filepath)
        mpart = MultipartEncoder(
            fields={'file': (filename, open(filepath, 'rb'), 'text/plain')}
        )
        response = requests.put(
            url,
            data=mpart,
            headers={
                'Content-Type': mpart.content_type,
                'Content-Length': '%d' % length,
            },
            auth=chacractl.config['credentials'],
            verify=chacractl.config['ssl_verify'])
        if response.status_code > 201:
            logger.warning("%s -> %s", response.status_code, response.text)
        # trim off binary filename
        url = url.rsplit('/', 2)[0] + "/"
        if not self.upload_is_verified(url, filename, digest):
            # Maybe the old file with a different digest is still there, so
            # don't delete it
            raise SystemExit(
                    'Checksum mismatch: server has wrong checksum for %s!'
                    % filepath)
    @retry()
    def delete(self, url):
        exists = requests.head(url, verify=chacractl.config['ssl_verify'])
        if exists.status_code == 404:
            logger.warning('resource already deleted')
            logger.warning('SKIP %s', url)
            return
        logger.info('DELETE file: %s', url)
        response = requests.delete(
            url,
            auth=chacractl.config['credentials'],
            verify=chacractl.config['ssl_verify'])
        if response.status_code < 200 or response.status_code > 299:
            logger.warning("%s -> %s", response.status_code, response.text)

    def main(self):
        self.parser = Transport(self.argv, options=self.options)
        self.parser.catch_help = self._help
        self.parser.parse_args()
        self.force = self.parser.has('--force')

        # handle posting binaries:
        if self.parser.has('create'):
            url_part = self.sanitize_url(self.parser.get('create'))
            if not sys.stdin.isatty():
                # read from stdin
                logger.info('reading input from stdin')
                for line in sys.stdin.readlines():
                    filename = self.sanitize_filename(line)
                    if not filename:
                        continue
                    url = os.path.join(self.base_url, url_part)
                    self.post(url, filename)
            else:
                filepath = self.sanitize_filename(self.argv[-1])
                if not filepath:
                    logger.warning(
                        'provided path does not exist: %s', self.argv[-1]
                    )
                    return
                url = os.path.join(self.base_url, url_part)
                self.post(url, filepath)

        elif self.parser.has('delete'):
            if self.parser.get('delete') is None:
                raise SystemExit('Specify a URL to delete a binary.')
            url_part = self.sanitize_url(self.parser.get('delete'))
            url = os.path.join(self.base_url, url_part)
            self.delete(url)
